src/seals/diagnostics/cliff_world.py

```diff
@@ -0,0 +1,142 @@
+"""A cliff world that uses the TabularModelPOMDP."""
+
+import numpy as np
+
+from seals.base_envs import TabularModelPOMDP
+
+
+class CliffWorldEnv(TabularModelPOMDP):
+    """A grid world with a goal next to a cliff the agent may fall into.
+
+    Illustration::
+
+         0 1 2 3 4 5 6 7 8 9
+        +-+-+-+-+-+-+-+-+-+-+  Wind:
+      0 |S|C|C|C|C|C|C|C|C|G|
+        +-+-+-+-+-+-+-+-+-+-+  ^ ^ ^
+      1 | | | | | | | | | | |  | | |
+        +-+-+-+-+-+-+-+-+-+-+
+      2 | | | | | | | | | | |  ^ ^ ^
+        +-+-+-+-+-+-+-+-+-+-+  | | |
+
+    Aim is to get from S to G. The G square has reward +10, the C squares
+    ("cliff") have reward -10, and all other squares have reward -1. Agent can
+    move in all directions (except through walls), but there is 30% chance that
+    they will be blown upwards by one more unit than intended due to wind.
+    Optimal policy is to go out a bit and avoid the cliff, but still hit goal
+    eventually.
+    """
+
+    width: int
+    height: int
+
+    def __init__(
+        self,
+        *,
+        width: int,
+        height: int,
+        horizon: int,
+        use_xy_obs: bool,
+        rew_default: int = -1,
+        rew_goal: int = 10,
+        rew_cliff: int = -10,
+        fail_p: float = 0.3,
+    ):
+        """Builds CliffWorld with specified dimensions and reward."""
+        assert (
+            width >= 3 and height >= 2
+        ), "degenerate grid world requested; is this a bug?"
+        self.width = width
+        self.height = height
+        succ_p = 1 - fail_p
+        n_states = width * height
+        O_mat = np.zeros(
+            (n_states, 2 if use_xy_obs else n_states),
+            dtype=np.float32,
+        )
+        R_vec = np.zeros((n_states,))
+        T_mat = np.zeros((n_states, 4, n_states))
+
+        # TODO(juan) this is not very elegant.
```

Comment:
What code is this referring to? `to_id_clamp` clamps `(x,y)` to legal points in the grid, then converts the pair to a scalar in a row-major format.